### DISCUSSION TITLE
rescale - what is it doing?

### DISCUSSION THREAD

**Discussion by mdavid800** [developmentseed/titiler#494] [Q&A] [answered]
HI Guys , 

Hopefully this isnt a daft question but i am trying to understand what Titiler is doing so i dont make a mistake. 

I have a COG data ser of Wind speed and it is in Float32 . In order to display this correctly i need to use the rescale function . However , i was hoping someone could confirm my understanding is correct . 

My data set ranges from 4.9m/s - 16.4m/s , when I use rescale am i correct in assuming that the data range min - max is split between 0-255 ? i.e. 0 represents 4.9m/s and 255 represents 16.4m/s ? 

if the above is correct,  is its possible to set the colormap such that the anything above say 10m/s is at the top of the colour scale ? I.e. from a wind energy perspective 16m/s is an extreme and i don't want my colour map scaled according to this outlier .

code for my get request is below. 

`r = httpx.get(
    f"{titiler_endpoint}/cog/tilejson.json",
    params = {
        "url": url,
        "rescale": f"{minv},{maxv}",
        "colormap_name": "viridis"
    }
).json()`

**Comment by geospatial-jeff** [comment, marked as answer]
Rescale is doing a linear stretch, the `minv` and `maxv` parameters in your GET request determine the input range for this stretch while the output range will always be 0-255 (8bit).  It's ultimately calling [this function](https://github.com/cogeotiff/rio-tiler/blob/62b0371836b0cb116a53459b5a79fcdc84f1cab1/rio_tiler/utils.py#L358-L378) in rio-tiler (note the call to `numpy.clip`).

> My data set ranges from 4.9m/s - 16.4m/s , when I use rescale am i correct in assuming that the data range min - max is split between 0-255 ? i.e. 0 represents 4.9m/s and 255 represents 16.4m/s ?

This is correct if you don't specify the `rescale` query parameter, in which case titiler falls back to scaling your data based on it's min/max range.  If you wanted to cut off the histogram at 10m/s to remove outliers you could set the `rescale` query parameter to `4.9,10`.

**Reply by mdavid800** [reply]
Thanks @geospatial-jeff  that helps me understand it. 

I assume if I set the rescale to 4.9 -10 m/s , everything above 10m/s would be set to 10m/s . 

If I wanted to keep the outliers but just have the values > 10m/s coloured the same could I do that in the colour map?

**Reply by geospatial-jeff** [reply]
Yes I believe you would have to define a custom colormap that classifies values above 10 m/s with the same color.  This would preserve the original values instead of clipping the image to the min/max values passed to `rescale`.  You can pass a custom JSON encoded colormap to the `colormap` query parameter.  Personally I've never done this before, so I'm not too familiar with the syntax.

**Comment by vincentsarago** [comment]
thanks @geospatial-jeff 🙏 

@mdavid80, as Jeff mentioned you can either use `rescale` to rescale your input data from {min}/{max} to 0->255 and then apply a linear colormap OR you can directly use `interval colormap` https://github.com/developmentseed/titiler/blob/master/src/titiler/core/tests/test_factories.py#L136-L148
